File: content/snippets/cpp/exercises/advanced/graph_traversal.cpp
#include <bits/stdc++.h>
using namespace std;
int main(){
    vector<vector<int>> g = {{1,2},{0,3},{0,3},{1,2}}; // simple graph
    // BFS
    vector<int> vis(4); queue<int>q; q.push(0); vis[0]=1;
    cout<<"BFS:";
    while(!q.empty()){int u=q.front();q.pop(); cout<<" "<<u; for(int v:g[u]) if(!vis[v]){vis[v]=1; q.push(v);} }
    cout<<"\n";
    // DFS
    vector<int> seen(4); function<void(int)> dfs=[&](int u){ seen[u]=1; cout<<u<<" "; for(int v:g[u]) if(!seen[v]) dfs(v); };
    cout<<"DFS: "; dfs(0); cout<<"\n"; }

File: content/snippets/typescript/exercises/advanced/graph_traversal.ts
const g: number[][] = [[1,2],[0,3],[0,3],[1,2]]
console.log('BFS:');
{
  const q = [0]; const seen = new Set([0]);
  while(q.length){ const u = q.shift()!; process.stdout.write(' '+u); for(const v of g[u]) if(!seen.has(v)){ seen.add(v); q.push(v); } }
  console.log();
}
console.log('DFS:');
{ const seen = new Set<number>(); function dfs(u:number){ seen.add(u); process.stdout.write(' '+u); for(const v of g[u]) if(!seen.has(v)) dfs(v); } dfs(0); console.log(); }
export {}
